display-servers/x11rb-display-server/src/xwrap/window.rs

```diff
@@ -0,0 +1,465 @@
+        // Gather info about the window from xlib.
+        let name = self.get_window_name(window)?;
+        let legacy_name = self.get_window_legacy_name(window)?;
+        let class = self.get_window_class(window)?;
+        let pid = self.get_window_pid(window)?;
+        let r#type = self.get_window_type(window)?;
+        let states = self.get_window_states(window)?;
+        let actions = self.get_window_actions_atoms(window)?;
+        let mut can_resize = actions.contains(&self.atoms.NetWMActionResize);
+        let trans = self.get_transient_for(window)?;
+        let sizing_hint = self.get_hint_sizing_as_xyhw(window)?;
+        let wm_hint = self.get_wmhints(window)?;
+
+        // Build the new window, and fill in info about it.
+        let mut w = Window::new(handle, Some(name), Some(pid));
+        w.res_name = class
+            .as_ref()
+            .and_then(|c| String::from_utf8(c.instance().to_vec()).ok());
+        w.res_class = class.and_then(|c| String::from_utf8(c.class().to_vec()).ok());
+        w.legacy_name = Some(legacy_name);
+        w.r#type = r#type.clone();
+        w.states = states;
+        w.transient = trans.map(|h| WindowHandle(X11rbWindowHandle(h)));
+        // // Initialise the windows floating with the pre-mapped settings.
+        // let xyhw = XyhwChange {
+        //     x: Some(attrs.x),
+        //     y: Some(attrs.y),
+        //     w: Some(attrs.width),
+        //     h: Some(attrs.height),
+        //     ..XyhwChange::default()
+        // };
+        // xyhw.update_window_floating(&mut w);
+        sizing_hint
+            .unwrap_or_default()
+            .update_window_floating(&mut w);
+        let mut requested = Xyhw::default();
+        // xyhw.update(&mut requested);
+        sizing_hint.unwrap_or_default().update(&mut requested);
+
+        if let Some(hint) = sizing_hint {
+            // Ignore this for now for non-splashes as it causes issues, e.g. mintstick is non-resizable but is too
+            // small, issue #614: https://github.com/leftwm/leftwm/issues/614.
+            can_resize = match (r#type, hint.minw, hint.minh, hint.maxw, hint.maxh) {
+                (
+                    WindowType::Splash,
+                    Some(min_width),
+                    Some(min_height),
+                    Some(max_width),
+                    Some(max_height),
+                ) => can_resize || min_width != max_width || min_height != max_height,
+                _ => true,
+            };
+            // Use the pre-mapped sizes if they are bigger.
+            // hint.w = std::cmp::max(xyhw.w, hint.w);
+            // hint.h = std::cmp::max(xyhw.h, hint.h);
+            hint.update_window_floating(&mut w);
+            hint.update(&mut requested);
+        }
+
+        w.requested = Some(requested);
+        w.can_resize = can_resize;
+        if let Some(hint) = wm_hint {
+            w.never_focus = hint.input.unwrap_or(false);
```

Comment:
@Syudagye does `hint.input == Some(true)` mean the window actually **wants** to be focused? i.e. did you forget to invert this flag?

Reply:
@Syudagye did you overlook this one?

I think this flag needs to be inverted i.e. you're currently trying to set if the window is **allowed to be focused** instead of it being **denied focus** (never_focus)

This table (and code) shows how I think this flag should be used:
| hint.input | never_focus |
|-|-|
| Some(true) | false |
| Some(false)| true |
| None | false |
```suggestion
            w.never_focus = !hint.input.unwrap_or(true);
```

Reference: https://tronche.com/gui/x/icccm/sec-4.html#s-4.1.7